### PR TITLE
refactor parameter on consumer url logic code

### PR DESCRIPTION
**Pull request by cheese8**
refactor parameter on consumer url logic code: #10312
1.using ParameterOnConsumerUrl, getParameterOnConsumerUrl and hasParameterOnConsumerUrl to refactor the multi-places similar code
2. using final

### CODE REVIEW

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java [hunk 1]
```diff
@@ -343,37 +312,91 @@ public boolean hasMethodParameter(String method, String key) {
...
+                return false;
+        }
+    }
+    
+    enum ParameterOnConsumerUrl {
+        Parameter,
+        ServiceParameter,
+        ServiceMethodParameter,
+        MethodParameter
+    }
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java:399]: Will this make code more unreadable

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java [hunk 2]
```diff
@@ -343,37 +312,91 @@ public boolean hasMethodParameter(String method, String key) {
...
         }
-
         String protocolServiceKey = getProtocolServiceKey();
         if (isEmpty(protocolServiceKey)) {
             return false;
         }
         return hasServiceMethodParameter(protocolServiceKey, method);
     }
+    
+    private String getParameterOnConsumerUrl(final ParameterOnConsumerUrl parameterOnConsumerUrl, final URL consumerUrl,
```
Comment (EarthChen) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java:338]: replace `if return` is better.

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java (hunk 1 above)
Comment (AlbumenJ) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java:399]: just replaced `consumerUrl` operation with `Function<URL, String>`

Comment (cheese8) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java]: As `using Function to refactor the switch` way?

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java (hunk 2 above)
Comment (cheese8) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java]: How about this way? As the code in `using Function to refactor the switch`.

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java (hunk 1 above)
Comment (AlbumenJ) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java]:
```
    public Optional<String> executeWithConsumerUrl (Function<URL, String> function) {
            URL consumerUrl = RpcContext.getServiceContext().getConsumerUrl();
            if (consumerUrl != null) {
                String v = function.apply(consumerUrl);
                if (StringUtils.isNotEmpty(v)) {
                    return Optional.of(v);
                }
                return Optional.empty();
            }
    }
```

**Comment by chickenlj** [comment]
Thanks for the contribution.

This code seems to be more maintenance-friendly but harder to read, considering the fact that the logic in InstanceAddressURL is very likely to be changed in the future. I would prefer to keep the current code unchanged.

**Pull request closed without merging**